drivers/adc/adc_ambiq.c

```diff
@@ -0,0 +1,389 @@
+	}
+
+	/* Set up an ADC slot */
+	ADCSlotConfig.eMeasToAvg = AM_HAL_ADC_SLOT_AVG_1;
+	ADCSlotConfig.eChannel = channel;
+	ADCSlotConfig.bWindowCompare = false;
+	ADCSlotConfig.bEnabled = true;
+	if (AM_HAL_STATUS_SUCCESS !=
+	    am_hal_adc_configure_slot(data->adcHandle, ui32SlotNumber, &ADCSlotConfig)) {
+		LOG_ERR("configuring ADC Slot 0 failed.\n");
+		return -ENODEV;
+	}
+
+	return 0;
+}
+
+static void adc_ambiq_isr(const struct device *dev)
+{
+	struct adc_ambiq_data *data = dev->data;
+	uint32_t ui32IntMask;
+	uint32_t ui32NumSamples;
+	am_hal_adc_sample_t Sample;
+
+	/* Read the interrupt status. */
+	am_hal_adc_interrupt_status(data->adcHandle, &ui32IntMask, true);
+	/* Clear the ADC interrupt.*/
+	am_hal_adc_interrupt_clear(data->adcHandle, ui32IntMask);
+
+	/*
+	 * If we got a conversion completion interrupt (which should be our only
+	 * ADC interrupt), go ahead and read the data.
+	 */
+	if (ui32IntMask & AM_HAL_ADC_INT_CNVCMP) {
+		for (uint32_t i = 0; i < data->active_channels; i++) {
+			/* Read the value from the FIFO. */
+			ui32NumSamples = 1;
+			am_hal_adc_samples_read(data->adcHandle, false, NULL, &ui32NumSamples,
+						&Sample);
+			*data->buffer++ = Sample.ui32Sample;
+		}
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+static int adc_ambiq_check_buffer_size(const struct adc_sequence *sequence, uint8_t active_channels)
+{
+	size_t needed_buffer_size;
+
+	needed_buffer_size = active_channels * sizeof(uint16_t);
+
+	if (sequence->options) {
+		needed_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed_buffer_size) {
+		LOG_DBG("Provided buffer is too small (%u/%u)", sequence->buffer_size,
+			needed_buffer_size);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_ambiq_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_ambiq_data *data = dev->data;
+	const struct adc_ambiq_config *cfg = dev->config;
+	uint8_t channel_id = 0;
+	uint32_t channels = 0;
+	uint8_t active_channels = 0;
+	uint8_t slot_index;
+
+	int error = 0;
+
+	if (sequence->channels & ~BIT_MASK(cfg->num_channels)) {
+		LOG_ERR("Incorrect channels, bitmask 0x%x", sequence->channels);
+		return -EINVAL;
+	}
+
+	if (sequence->channels == 0UL) {
+		LOG_ERR("No channel selected");
+		return -EINVAL;
+	}
+	active_channels = POPCOUNT(sequence->channels);
+	if (active_channels > AMBIQ_ADC_SLOT_BUMBER) {
+		LOG_ERR("Too many channels for sequencer. Max: %d", AMBIQ_ADC_SLOT_BUMBER);
+		return -ENOTSUP;
+	}
+
+	channels = sequence->channels;
+	for (slot_index = 0; slot_index < active_channels; slot_index++) {
+		channel_id = find_lsb_set(channels) - 1;
+		error = adc_ambiq_slot_config(dev, sequence, channel_id, slot_index);
+		if (error < 0) {
+			return error;
+		}
+		channels &= ~BIT(channel_id);
+	}
+	__ASSERT_NO_MSG(channels == 0);
+
+	/* Enable the ADC. */
+	am_hal_adc_enable(data->adcHandle);
+
+	error = adc_ambiq_check_buffer_size(sequence, active_channels);
```

Comment:
It would be probably better to check the buffer size before configuring the slots and enabling the ADC.